File: 81. Search in Rotated Sorted Array II.java
class Solution {
    public boolean search(int[] nums, int target) {
        int lo = 0;
        int hi = nums.length - 1;
        while (lo <= hi) {
            int mid = lo + (hi - lo)/2;
            if (nums[mid] == target) return true;
            else if (nums[mid] > target) {
                if (nums[mid] > nums[lo]) { // left side is sorted
                    if (target < nums[lo]) {lo = mid + 1;}
                    else {hi = mid - 1;}
                }
                else if (nums[mid] == nums[lo]) {
                    lo++;
                }
                else {// right side is sorted 
                   hi = mid - 1; 
                }
            }
            
            else { // nums[mid] < target
                if (nums[mid] < nums[hi]) { // right side is sorted
                    if (target > nums[hi]) {hi = mid - 1;}
                    else lo = mid + 1;
                }
                else if (nums[mid] == nums[hi]) {
                    hi--;
                }
                else {// left side is sorted, target is larger than mid
                    lo = mid + 1;
                }
            }
        }
        return false;
        
    }
}

// O(logn) worst case O(n)

File: amazon/python/81. Search in Rotated Sorted Array II.py
class Solution(object):
    def search(self, nums, target):
        """
        :type nums: List[int]
        :type target: int
        :rtype: bool
        """
        if len(nums) == 0:
            return False
        if len(nums) == 1:
            return target == nums[0]
        
        l = 0
        r = len(nums) - 1
        while l <= r:
            mid = l + (r - l)//2
            if nums[mid] == target:
                return True
            elif nums[mid] == nums[l]:
                l = l + 1
            elif nums[mid] > nums[l]:
                if target >= nums[l] and target < nums[mid]:
                    r = mid - 1
                else:
                    l = mid + 1
            else: # nums[mid] < nums[l]
                if target > nums[mid] and target <= nums[r]:
                    l = mid + 1
                else:
                    r = mid - 1
                    
        return False
                
    
    # Time: o(LOGN) => o(N)
    # Space: O(1)
